distributed/protocol.py

```diff
@@ -64,35 +63,89 @@
     default_compression = 'lz4'
 
 
+def extract_big_bytes(x):
+    big = {}
+    _extract_big_bytes(x, big)
+    if big:
+        x = deepcopy(x)
+        for path in big:
+            t = get_in(path[:-1], x)
+            if isinstance(t, dict):
+                del t[path[-1]]
+            else:
+                t[path[-1]] = None
+    return x, big
+
+
+def _extract_big_bytes(x, big, path=()):
+    if type(x) is dict:
+        for k, v in x.items():
+            if isinstance(v, (list, dict)):
+                _extract_big_bytes(v, big, path + (k,))
+            elif type(v) is bytes and len(v) >= 2**31:
+                big[path + (k,)] = v
+    elif type(x) is list:
+        for k, v in enumerate(x):
+            if isinstance(v, (list, dict)):
+                _extract_big_bytes(v, big, path + (k,))
+            elif type(v) is bytes and len(v) >= 2**31:
+                big[path + (k,)] = v
```

Comment:
Not quite. We crawl through `msg` and find all the big strings and paths - these are then put into `big`. After that, if `big` isn't empty, we `deepcopy` the original `msg`, and then remove all the big strings from the copy. Cheap in the common case when `big` is empty, but more expensive than just mutating `msg` originally.